src/container/srv_container.c

```diff
@@ -960,6 +971,265 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl,
+	ds_cont_child_put(cont_child);
+	return rc;
+}
+
+int
+ds_cont_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid, daos_epoch_t eph)
+{
+	struct refresh_vos_agg_eph_arg	arg;
+	int				rc;
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_uuid, cont_uuid);
+	arg.min_eph = eph;
+
+	rc = dss_task_collective(cont_refresh_vos_agg_eph_one, &arg, 0,
+				 DSS_ULT_IO);
+	return rc;
+}
+
+#define EC_AGG_EPH_INTV	 (10ULL * NSEC_PER_SEC)	/* seconds interval to check*/
+static void
+cont_agg_eph_leader_ult(void *arg)
+{
+	struct dss_sleep_ult	*sleep_ult;
+	struct cont_svc		*svc = arg;
+	int			rc = 0;
+
+	sleep_ult = dss_sleep_ult_create();
+	if (sleep_ult == NULL) {
+		D_ERROR(DF_UUID " sleep ult creation failed.\n",
+			DP_UUID(svc->cs_pool->sp_uuid));
+		return;
+	}
+
+	while(1) {
+		struct cont_ec_agg *ec_agg;
+
+		d_list_for_each_entry(ec_agg, &svc->cs_ec_agg_list,
+				      ea_list) {
+			daos_epoch_t min_eph = DAOS_EPOCH_MAX;
+			int	     i;
+
+			if (svc->cs_pool->sp_stopping)
+				break;
+
+			for (i = 0; i < ec_agg->ea_servers_num; i++) {
+				if (ec_agg->ea_server_ephs[i].eph < min_eph)
+					min_eph = ec_agg->ea_server_ephs[i].eph;
+			}
+
+			if (min_eph <= ec_agg->ea_current_eph)
+				continue;
+
+			D_DEBUG(DB_MD, DF_CONT" sync "DF_U64"\n",
+				DP_CONT(svc->cs_pool_uuid, ec_agg->ea_cont_uuid),
+				min_eph);
+			rc = cont_iv_ec_agg_eph_refresh(
+						svc->cs_pool->sp_iv_ns,
+						ec_agg->ea_cont_uuid, min_eph);
+			if (rc) {
+				D_ERROR(DF_CONT": refresh failed: %d\n",
+					DP_CONT(svc->cs_pool->sp_uuid,
+						ec_agg->ea_cont_uuid), rc);
+				continue;
+			}
+			ec_agg->ea_current_eph = min_eph;
+		}
+
+		if (svc->cs_ec_agg_stopping)
+			break;
+		dss_ult_sleep(sleep_ult, EC_AGG_EPH_INTV);
+	}
+
+	D_DEBUG(DF_DSMS, DF_UUID": stop eph ult: rc %d\n",
+		DP_UUID(svc->cs_pool_uuid), rc);
+
+	dss_sleep_ult_destroy(sleep_ult);
+	ABT_mutex_lock(svc->cs_ec_agg_stop_mutex);
+	ABT_cond_signal(svc->cs_ec_agg_stop_cond);
+	ABT_mutex_unlock(svc->cs_ec_agg_stop_mutex);
+}
+
+static void
+cont_svc_ec_agg_destory(struct cont_svc *svc)
+{
+	struct cont_ec_agg *ec_agg;
+	struct cont_ec_agg *tmp;
+
+	d_list_for_each_entry_safe(ec_agg, tmp, &svc->cs_ec_agg_list, ea_list) {
+		d_list_del(&ec_agg->ea_list);
+		if(ec_agg->ea_server_ephs)
+			D_FREE(ec_agg->ea_server_ephs);
+		D_FREE(ec_agg);
+	}
+
+	if (svc->cs_ec_agg_stop_cond)
+		ABT_cond_free(&svc->cs_ec_agg_stop_cond);
+
+	if (svc->cs_ec_agg_stop_mutex)
+		ABT_mutex_free(&svc->cs_ec_agg_stop_mutex);
+}
+
+static int
+cont_svc_ec_agg_leader_start(struct cont_svc *svc)
+{
+	int	rc;
+
+	D_INIT_LIST_HEAD(&svc->cs_ec_agg_list);
+	rc = ABT_cond_create(&svc->cs_ec_agg_stop_cond);
+	if (rc != ABT_SUCCESS)
+		D_GOTO(out, rc = dss_abterr2der(rc));
+
+	rc = ABT_mutex_create(&svc->cs_ec_agg_stop_mutex);
+	if (rc != ABT_SUCCESS)
+		D_GOTO(out, rc = dss_abterr2der(rc));
+
+	rc = dss_ult_create(cont_agg_eph_leader_ult, svc, DSS_ULT_MISC,
+			    DSS_TGT_SELF, 0, NULL);
```

Comment:
DSS_ULT_MISC is scheduled on helper xstream, is that what you want?

Reply:
this needs to be on stream 0. so TGT_SELF.